### PR TITLE
feat(HMS-761): add instance description to aws job

### PR DESCRIPTION
**Pull request by amirfefer**
This adda fetching instance description to aws job, stores the instance's public ip and dns into reservation_instances table

I keep this PR as draft until UI is ready

### CODE REVIEW

File: internal/jobs/launch_instance_aws.go
Comment (ezr-ondrej) [inline internal/jobs/launch_instance_aws.go:64]: Can we move this to the above statement so it is clear we are checking the error of that statement? It's common practice not to divide by new line for these cases.

OTH I'd add new line after this if, as the following statement is not related :)

File: internal/jobs/launch_instance_aws.go
Comment (ezr-ondrej) [inline internal/jobs/launch_instance_aws.go:252]: This forces the 6 second wait, shall we instead wait a second, check if ready and repeat six times?
I believe usually AWS is much faster, so there is not much reason to wait 6 seconds for all reservations, or is there some research you've done how long it takes?

File: internal/clients/http/ec2/ec2_client.go
Comment (ezr-ondrej) [inline internal/clients/http/ec2/ec2_client.go:443]:
```suggestion
		return nil, errors.New("no reservation found in AWS response")
```

Comment (ezr-ondrej) [reply internal/clients/http/ec2/ec2_client.go:443]: Also as linter says, this error should be a constant variable.

**Review by ezr-ondrej — COMMENTED** [review]
Great, this looks on the right track, I'd really love if we could add some polling instead of static wait time tho...

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [inline internal/clients/http/ec2/ec2_client.go:341]: I would probably keep the naming convention from the AWS API here, even if it returns a slice. The call is named "DescribeInstances" so we could have the same name in our own interface too. Something like "DescribeInstanceDetails" maybe? (I like "Details" more than "Description" so the struct type could be `InstanceDetail`)

File: internal/clients/instance_description.go
Comment (lzap) [inline internal/clients/instance_description.go:13]: Can we not use string pointers? Pointers are bad for strings (string is already a pointer itself) and the only reason AWS SDK use them is because the types are generated from XML. Just use strings and empty value means "nothing". You will have to convert them, we have a package `ptr` with helper functions for that.

File: internal/clients/http/ec2/ec2_client.go
Comment (lzap) [reply internal/clients/http/ec2/ec2_client.go:341]: Another idea for the name as I commented down below: `ReservationInstance`

**Comment by lzap** [comment]
Dropped some comments, I think we should use JSONB for these fields as they are only "informative" and very likely not something we would need to create database index on. We do the same for AWS details.

File: internal/migrations/sql/018_add_description_reservation_instances.sql
Comment (lzap) [reply internal/migrations/sql/018_add_description_reservation_instances.sql:5]: I forgot the yaml tags I think these are needed for OpenAPI, but you get the idea.

File: internal/jobs/launch_instance_aws.go
Comment (avitova) [reply internal/jobs/launch_instance_aws.go:252]: Also, there we are using the first page of the DescribeInstanceOutput, and only the [first item](https://github.com/RHEnVision/provisioning-backend/blob/67fef988a14ec93adf037e27eef77cada6d1f32d/internal/clients/http/ec2/ec2_client.go#L445) in the array.
There could be something in the NextToken field. As well as the reservation does not have to be on this index in the reservation array?
I wonder, is it safe since there is at least 6-second gap between the launching and obtaining the description?

**Comment by lzap** [comment]
Just for the record, what am I suggesting is to store new data as JSON in the database in a single column with JSONB type. Postgres allows selecting data from JSON directly and even indexing, although we will never need to utilize indexing hopefully.

image

File: internal/jobs/launch_instance_aws.go
Comment (amirfefer) [reply internal/jobs/launch_instance_aws.go]: @avitova, as far as I understand, no need to iterate the rerservations array, because all of the instances are related to the same reservation, so only one reservation should return. 
Regarding pagination, I guess for now it is not needed, the max instances count for now is 45

**Comment by amirfefer** [comment]
Thanks @lzap , @ezr-ondrej , @avitova  for the review
Hopefully I've addressed them all :)

File: internal/jobs/launch_instance_aws.go
Comment (lzap) [inline internal/jobs/launch_instance_aws.go:267]: Note for me or @ezr-ondrej - the https://github.com/RHEnVision/provisioning-backend/pull/461 requires all functions to use the new `nilUnlessTimeout` here so if this gets merged before this needs to be updated.

File: internal/migrations/sql/019_add_detail_reservation_instances.sql
Comment (lzap) [inline internal/migrations/sql/019_add_detail_reservation_instances.sql:2]: We do just `detail JSONB` which is nullable with no default value in the other table but this is fine.

**Comment by lzap** [comment]
It’s perfect, going to test it now.

**Comment by lzap** [comment]
Two problems before I can start testing tho revealed by our CI:

* You need to regenerate OpenAPI since you are adding new fields to the type.
* IQE now fails with `iqe_provisioning_api.exceptions.ApiTypeError: Invalid type for variable '0'. Required value type is str and passed type was dict at ['received_data']['instances'][0]`

This is because you are changing a type of existing field. I think it is fine since we are not V1 API yet, opinions @ezr-ondrej @akhil-jha ?

File: internal/payloads/reservation_payload.go
Comment (lzap) [inline internal/payloads/reservation_payload.go:84]: FYI this is what is causing IQE test break, this is a reasonable change so my proposal is to fix OpenAPI and wait for @akhil-jha to prepare IQE PR so we can merge this at once.

Comment (akhil-jha) [reply internal/payloads/reservation_payload.go:84]: Sure, tag me in the updated PR

**Comment by lzap** [comment]
Looks like OpenAPI is now ready, @akhil-jha thanks!

File: internal/payloads/reservation_payload.go
Comment (lzap) [inline internal/payloads/reservation_payload.go:47]: Nitpick: since this is named "instances" in JSON this could be also named "Instance".

File: internal/jobs/launch_instance_aws.go
Comment (amirfefer) [inline internal/jobs/launch_instance_aws.go:264]: forgot the `break`... now it much faster :)

**Review by ezr-ondrej — APPROVED** [review]
Changing the OpenAPI field is fine IMHO :+1:

File: internal/jobs/launch_instance_aws.go
Comment (lzap) [reply internal/jobs/launch_instance_aws.go:264]: Quite honestly, now that you brought my attention, this for-continue-break is something that I see maybe for the first time in my life. I would always write:

```
for ... {
  // request
  if there is a response {
    // update and break out
  }
}
```

This is more natural use of for loop which by design - loops so there is no need to do continue on each cycle. What you have is little bit more clunky, it is acceptable I guess :-)

File: internal/payloads/reservation_payload.go
Comment (lzap) [inline internal/payloads/reservation_payload.go:47]: Please do not use embedding here since it also pulls `instance_id` unwanted field (makes no sense to expose it through the API). Just copy the fields and in the copy function copy one by one.

**Review by lzap — CHANGES_REQUESTED** [review]
I need to ask for one small change because OpenAPI now exposes an unwanted field.

Thanks.

File: internal/payloads/reservation_payload.go
Comment (amirfefer) [reply internal/payloads/reservation_payload.go:47]: why the instance_id is unwanted?

File: internal/jobs/launch_instance_aws.go
Comment (amirfefer) [reply internal/jobs/launch_instance_aws.go]: I agree, it's more clunky :)
I removed the `continue` and using just the break now

File: internal/payloads/reservation_payload.go
Comment (ezr-ondrej) [reply internal/payloads/reservation_payload.go:47]: I'd agree with amir here, we are exposing the instance_id in the UI, so we should include it.
I'd back the copying is bit better because its bit more clear what data are we exposing.
I believe the field we don't want is `reservation_id`, as that's already included in the response.

**Review by ezr-ondrej — APPROVED** [review]
LGTM, I'd probably not nest the dns and ipv4 in details, but that's just my preference, this is fine.
Great job @amirfefer :+1:

**Review by lzap — APPROVED** [review]
Perfect, thank you! Very nice job.

**Pull request merged**
